corgidrp/l3_to_l4.py

```diff
@@ -1078,7 +1079,317 @@ def subtract_stellar_polarization(input_dataset, system_mueller_matrix_cal, nd_m
+                        )
+            frame.err[0,0,:,:] = np.sqrt(sum_err**2 + diff_err**2) / 2
+            frame.err[0,1,:,:] = frame.err[0,0,:,:]
+            
+            updated_frames.append(frame)
+
+    updated_dataset = data.Dataset(updated_frames)
+    history_msg = f"Subtracted Apparent Stellar Polarization, stellar Q value: {S_in[1]}, stellar Q err: {S_in_err[1]}, \
+    stellar U value: {S_in[2]}, stellar U err: {S_in_err[2]}."
+    updated_dataset.update_after_processing_step(history_msg)
+    return updated_dataset
+
+def spec_psf_subtraction(input_dataset):
+    '''
+    RDI PSF subtraction for spectroscopy mode.
+    Assumes the reference images are marked with PSFREF=True in the primary header
+    and that they all have the same alignment.
+
+    Args:
+        input_dataset (corgidrp.data.Dataset): L3 dataset containing the science and reference images
+    
+    Returns:
+        corgidrp.data.Dataset: dataset containing the PSF-subtracted science images and the mean reference image as the last image
+    
+    '''
+    dataset = input_dataset.copy()
+    input_datasets, values = dataset.split_dataset(prihdr_keywords=["PSFREF"])
+    if values != [0,1] and values != [1,0]:
+        raise ValueError("PSFREF keyword must be present in the primary header and be either 0 or 1 for all images")
+    ref_index = values.index(True)
+    mean_ref_dset = combine_subexposures(input_datasets[ref_index], num_frames_per_group=None, collapse="mean", num_frames_scaling=False)
+    # undo any NaN assignments in the image since we FFT below
+    nan_inds = np.where(np.isnan(mean_ref_dset[0].data))
+    if len(nan_inds[0]) > 0:
+        mean_ref_dset[0].data[nan_inds] = np.mean(input_datasets[ref_index].all_data[:,nan_inds[0],nan_inds[1]], axis=0)
+    mean_ref = mean_ref_dset[0].copy()
+    all_data = []
+    all_dq = []
+    all_err = []
+    image_list = []
+    for frame in input_datasets[1-ref_index]:    
+        # compute shift between frame and mean_ref 
+        shift = get_shift_correlation(frame.data, mean_ref.data)
+        # shift mean_ref to be on top of frame data
+        shifted_ref = np.roll(mean_ref.data, (shift[0], shift[1]), axis=(0,1))
+        # rescale wavelengh bands to match
+        ref_col_mean = np.mean(shifted_ref,axis=0)
+        ref_col_mean[ref_col_mean==0] = 1 # prevent div by 0
+        scale = np.mean(frame.data,axis=0)/ref_col_mean
+        shifted_scaled_ref = shifted_ref*scale
+
+        shifted_refdq = np.roll(mean_ref.dq, (shift[0], shift[1]), axis=(0,1))
+        # at this point in the pipeline, the err is mainly shot noise, so multiplying the err is appropriate
+        # shifting may throw off err at the edges of the frame, but those pixels aren't used anyways
+        shifted_scaled_referr = np.roll(mean_ref.err[0], (shift[0], shift[1]), axis=(0,1))*scale
+        # to prevent affecting future steps with error propagated forward from pixels here we don't use: 
```

Comment:
I don't understand the desire to set the rest of the error map to 0's. Does it do something bad if we don't do this step? in particular, the exact choice of crop region seem arbitrary and I'd rather things be uniform.